mongomock/collection.py

```diff
@@ -1754,15 +1754,40 @@ def _extend_collection(out_collection, field, expression):
                                 unwound_collection[-1], v[1:], field_item)
                     out_collection = unwound_collection
                 elif k == '$project':
-                    filter_list = ['_id']
+                    filter_list = []
+                    method = None
+                    include_id = v.get('_id')
+
                     for field, value in iteritems(v):
-                        if field == '_id' and not value:
-                            filter_list.remove('_id')
-                        elif value:
-                            filter_list.append(field)
+                        if field == '_id':
+                            continue
+                        if method is None:
+                            method = 'include' if value else 'exclude'
+                            if method == 'exclude' and include_id:
+                                raise ValueError(
+                                    "Bad projection specification, cannot include fields "
+                                    "or add computed fields during an exclusion projection: %s" % v)
+                        elif method == 'include' and not value and field != '_id':
```

Comment:
No need to check for field not being _id as you would already skip the whole loop (see `continue` line 1763).